File: Top_Interview_150/Matrix/python/73.py
from typing import List


class Solution:

    @staticmethod
    def set_null_line(matrix, x, m):
        for i in range(m):
            matrix[x][i] = 0

    @staticmethod
    def set_null_column(matrix, y, n):
        for i in range(n):
            matrix[i][y] = 0

    def setZeroes(self, matrix: List[List[int]]) -> None:

        n = len(matrix)
        m = len(matrix[0])
        null_lines = set()
        null_columns = set()

        for i in range(n):
            for j in range(m):
                if matrix[i][j] == 0:
                    null_lines.add(i)
                    null_columns.add(j)

        for x in null_lines:
            self.set_null_line(matrix, x, m)

        for y in null_columns:
            self.set_null_column(matrix, y, n)

File: Top_Interview_150/Matrix/cpp/73.cpp
#include <vector>
#include <unordered_set>

using namespace std;

class Solution
{

    void set_null_line(vector<vector<int>> &matrix, int x, int m)
    {
        for (int i = 0; i < m; ++i)
            matrix[x][i] = 0;
    }

    void set_null_column(vector<vector<int>> &matrix, int y, int n)
    {
        for (int i = 0; i < n; ++i)
            matrix[i][y] = 0;
    }

public:
    void setZeroes(vector<vector<int>> &matrix)
    {
        int n = matrix.size();
        int m = matrix[0].size();
        unordered_set<int> null_lines;
        unordered_set<int> null_columns;

        for (int i = 0; i < n; ++i)
        {
            for (int j = 0; j < m; ++j)
            {
                if (matrix[i][j] == 0)
                {
                    null_lines.insert(i);
                    null_columns.insert(j);
                }
            }
        }

        for (int x : null_lines)
            this->set_null_line(matrix, x, m);

        for (int y : null_columns)
            this->set_null_column(matrix, y, n);
    }
};